File: 20_Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for sym in s:
            stack.append(sym)
            if len(stack) >= 2:
                for_check = "".join(stack[-2:])
                if for_check in ["()", "{}", "[]"]:
                    stack = stack[:-2]

        return not stack


sol = Solution()
print(sol.isValid("()"))
print(sol.isValid("())"))
print(sol.isValid("[()]"))
print(sol.isValid("[[[[[[[[({{}}})]]]]]]]]"))

File: 20_Valid_Parentheses.go
package main

func isValid(s string) bool {
	stack := []rune{}
	for _, sym := range s {
		switch sym {
		case '(', '[', '{':
			stack = append(stack, sym)
		default:
			if length := len(stack); length == 0 ||
				(sym == ')' && stack[length-1] != '(') ||
				(sym == '}' && stack[length-1] != '{') ||
				(sym == ']' && stack[length-1] != '[') {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}
